### ISSUE TITLE
gsutil cp hangs on many small files when running in parallel

### ISSUE THREAD

**Issue by johanhaleby** [GoogleCloudPlatform/gsutil#464]
I have a repository that GCS bucket millions of small files in different folders. When I run:

```bash
$ gsutil -m cp -r gs://my-bucket .
```
The process with eventually hang before completion, sometimes after 5 minutes and sometimes after several hours. This seems to be is 100% reproducible. I'm using version 4.27 but this has happened in older versions as well. As a workaround have I to use:

```bash
$ gsutil cp -r gs://my-bucket .
```

which works but it takes several days to download everything so it's not optimal.

**Comment by houglum** [comment]
Is there **any** chance you're running another instance of gsutil separately while that one is still working?  This is the only thing I've found that will almost certainly make gsutil hang (due to the separate gsutil processes/threads trying to use the same directory, `"$HOME/.gsutil"` unless otherwise specified, to manage state and "locked" auth cache files).

If you're not sure, I'd recommend invoking that command with the option to tell gsutil to use a different, non-default state dir, e.g.:

    gsutil -o "GSUtil:state_dir=/absolute/path/to/some/directory" -m cp -r gs://my-bucket .

(and don't have any other invocations of gsutil specify that directory).  If that still hangs, we'll know the state directory isn't the problem... at which point, you might want to run the command with the top-level `-D` option and see if the debug logs reveal anything fishy, or even show at what point gsutil stopped doing anything useful.

**Comment by AdamStelmaszczyk** [comment]
I reproduced the hanging with `for i in {1..1000}; do echo "gsutil -m rm gs://your_bucket/test/$i"; done | parallel`. It's fine if there are no files in `your_bucket/test`, it will still hang. In the other session one can check `ps x | wc -l` for the number of running processes, this number will fluctuate, but after about 2 minutes it will stay constant at pretty high number.

```
ps aux --forest | more
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      912256  0.0  0.0  18216  3400 ?        Ss   09:45   0:00 bash
root     1767600  0.0  0.0  36636  2904 ?        R+   10:02   0:00  \_ ps aux --forest
root     1767601  0.0  0.0   6420   872 ?        S+   10:02   0:00  \_ more
root        1085  0.1  0.0  38052 23024 ?        Ss   08:50   0:06 bash
root     1100123  3.3  0.0  45244 16900 ?        S+   09:58   0:07  \_ perl /usr/local/bin/parallel
root     1135909  0.0  0.0   4280   752 ?        S    09:58   0:00      \_ /bin/sh /usr/bin/gsutil -m rm gs://your_bucket/test/57
root     1136093  0.1  0.0  71644 27420 ?        S    09:58   0:00          \_ python2 /usr/bin/../lib/google-cloud-sdk/bin/bootstrapping/gsutil.py -m rm gs://your_bucket/test/57
root     1136707  0.2  0.0 311592 41280 ?        S    09:58   0:00              \_ /usr/bin/python2 /usr/lib/google-cloud-sdk/
platform/gsutil/gsutil -o GSUtil:default_project_id=your_project_id -m rm your_bucket/test/57
root     1142178  0.0  0.0      0     0 ?        Z    09:58   0:00                  \_ [python2] <defunct>
root     1142194  0.0  0.0      0     0 ?        Z    09:58   0:00                  \_ [python2] <defunct>
root     1142202  0.0  0.0      0     0 ?        Z    09:58   0:00                  \_ [python2] <defunct>
root     1142231  0.0  0.0      0     0 ?        Z    09:58   0:00                  \_ [python2] <defunct>
root     1142244  0.0  0.0 459056 36648 ?        Sl   09:58   0:00                  \_ /usr/bin/python2 /usr/lib/google-cloud-
sdk/platform/gsutil/gsutil -o GSUtil:default_project_id=your_project_id -m rm gs://your_bucket/test/57
root     1142275  0.0  0.0      0     0 ?        Z    09:58   0:00                  \_ [python2] <defunct>
root         768  0.0  0.0  18364  3460 ?        Ss+  Aug08   0:00 bash
root           1  0.0  0.0  17972  2896 ?        Ss   Aug08   0:00 /bin/bash /scripts/entrypoint.sh sleep 1000d
root          37  0.0  0.0   4188   708 ?        S    Aug08   0:00 sleep 1000d
root      909913  2.6  0.0 5896420 50788 ?       Sl   09:43   0:30 /usr/bin/python2 /usr/lib/google-cloud-sdk/platform/gsutil/
gsutil -o GSUtil:default_project_id=your_project_id -m cp -r gs://your_bucket/test .
```
Both `parallel` and `-m` seem crucial for reproduction, without any of them it didn't hang.

I tried reproducing with `gsutil cp`, but couldn't.

I tried providing unique `state_dir`: 
```
for i in {1..1000}; do mkdir /tmp/$i; done
for i in {1..1000}; do echo "gsutil -o "GSUtil:state_dir=/tmp/$1" -m rm gs://your_bucket/test/$i"; done | parallel
```

But it also hangs. Ran `gsutil` with `-D` flag but didn't see anything interesting. One weird thing is that it seems to look at files above 1000, don't know why (but not sure if related):
```
INFO 0812 10:38:01.681003 base_api.py] Calling method storage.objects.list with StorageObjectsListRequest: <StorageObjectsListRequest
 bucket: u'your_bucket'
 delimiter: u'/'
 maxResults: 1000
 prefix: u'test/999'
 projection: ProjectionValueValuesEnum(noAcl, 1)
 versions: False>
INFO 0812 10:38:01.681976 base_api.py] Making http GET to https://www.googleapis.com/storage/v1/b/your_bucket/o?prefix=test%2F999&projection=noAcl&versions=False&fields=prefixes%2CnextPageToken%2Citems%2Fname&alt=json&delimiter=%2F&maxResults=1000
INFO 0812 10:38:01.682480 base_api.py] Headers: {'accept': 'application/json',
 'accept-encoding': 'gzip, deflate',
 'content-length': '0',
 'user-agent': 'apitools gsutil/4.41 Python/2.7.13 (linux2) google-cloud-sdk/257.0.0 analytics/disabled'}
INFO 0812 10:38:01.682612 base_api.py] Body: (none)
INFO 0812 10:38:01.726975 base_api.py] Response of type Objects: <Objects
 items: [<Object
 acl: []
 name: u'test/9990'>, <Object
 acl: []
 name: u'test/9991'>, <Object
 acl: []
 name: u'test/9992'>, <Object
 acl: []
 name: u'test/9993'>, <Object
 acl: []
 name: u'test/9994'>, <Object
 acl: []
 name: u'test/9995'>, <Object
 acl: []
 name: u'test/9996'>, <Object
 acl: []
 name: u'test/9997'>, <Object
 acl: []
 name: u'test/9998'>, <Object
 acl: []
 name: u'test/9999'>]
 prefixes: []>
process count: 16
thread count: 5
DEBUG: Exception stack trace:
    Traceback (most recent call last):
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 626, in _RunNamedCommandAndHandleExceptions
        user_project=user_project)
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/command_runner.py", line 411, in RunNamedCommand
        return_code = command_inst.RunCommand()
      File "/usr/lib/google-cloud-sdk/platform/gsutil/gslib/commands/rm.py", line 386, in RunCommand
        (self.op_failure_count, plural_str, plural_str))
    CommandException: CommandException: 1 files/objects could not be removed.
    
CommandException: 1 files/objects could not be removed.
```

**Comment by dxdydz** [comment]
Just to add, seeing a hang with 4.48 on macOS using `cat files.txt | gsutil -m cp -I ./`

About 17k/100k files will copy and then it will hang.

**Comment by obrie** [comment]
For what it's worth, I found that using only threads for parallelization (and not child processes) appears to avoid the underlying deadlock here.  e.g. `-o GSUtil:parallel_process_count=1 -o GSUtil:parallel_thread_count=24`

**Comment by cdg720** [comment]
i also see this problem with gcloud version 310

**Comment by brendanreardon** [comment]
Are you all running these as subprocesses by chance? I had this issue when running gsutil through python with subprocess. Calling gsutil with a full path resolved the hanging for me. 

Specifically, my python code now looks like this example,

```python
def get_gsutil_path():
    cmd = "which gsutil"
    path = subprocess.check_output(cmd, shell=True, stderr=subprocess.PIPE)
    return path.decode("utf-8").replace("\n", "")

def glob_bucket(bucket):
    cmd = f'{get_gsutil_path()} ls gs://{bucket}/**'
    bucket_files = subprocess.check_output(cmd, shell=True, stderr=subprocess.PIPE)
    series = pd.read_csv(io.StringIO(bucket_files.decode('utf-8')), sep='\n', header=None).loc[:, 0]
    return series.tolist()
```

**Comment by dave-mclean** [comment]
I'm seeing the same issue. Downgrading to an earlier version (`297.0.1`) of gcloud has not "resolved it" for me though, the copy still hangs.

**Comment by saumyasuhagiya** [comment]
Facing the same issue even after updating to the latest version.

**Comment by cooolinx** [comment]
same issue here, using Python subprocess just like @brendanreardon do, any news?

**Comment by dilipped** [comment]
Are you using macOS? If yes, then this is a known issue for mac because of multiprocessing using fork. You can set parallel_process_count=1 to disable multiprocessing as mentioned here https://github.com/GoogleCloudPlatform/gsutil/issues/464#issuecomment-633334888

**Comment by cooolinx** [comment]
> Are you using macOS? If yes, then this is a known issue for mac because of multiprocessing using fork. You can set parallel_process_count=1 to disable multiprocessing as mentioned here [#464 (comment)](https://github.com/GoogleCloudPlatform/gsutil/issues/464#issuecomment-633334888)

Yes I am using macOS, thanks for the information, `parallel_process_count=1` did work

**Comment by JeremyKeustersML6** [comment]
Not sure if it is a coincidence or not, but updating the Google Cloud Components to the latest version (`Google Cloud SDK 342.0.0`) fixed this issue for me.

**Comment by wobushizhanghua** [comment]
> For what it's worth, I found that using only threads for parallelization (and not child processes) appears to avoid the underlying deadlock here. e.g. `-o GSUtil:parallel_process_count=1 -o GSUtil:parallel_thread_count=24`

Thanks, you saved my day

**Comment by jli** [comment]
Just gonna mention that GCP's Storage Transfer Service provides a way to transfer large amount of data between GCS buckets very quickly, which may be a workable alternative for some people's usecases.

**Comment by btc** [comment]
Google Cloud SDK 362.0.0
bq 2.0.71
core 2021.10.21
gsutil 5.4

Passing `-o GSUtil:parallel_process_count=1` to the `gsutil` command when using `-m` worked for me as well.

**Comment by astronautas** [comment]
> For what it's worth, I found that using only threads for parallelization (and not child processes) appears to avoid the underlying deadlock here. e.g. `-o GSUtil:parallel_process_count=1 -o GSUtil:parallel_thread_count=24`

Add this to your gsutil's .boto file.
Run `gsutil version -l` to see where it (.boto) resides.

**Comment by simplenotezy** [comment]
I tried with:

    gsutil -m -o GSUtil:parallel_process_count=1 -o GSUtil:parallel_thread_count=24 cp -n -r . gs://my-bucket

But to no avail. It still gets stuck unfortunately.

**Comment by btc** [comment]
@simplenotezy i am curious if it hangs if you omit the thread count flag.

**Comment by shaneiseminger** [comment]
Found this thread as I was experiencing the same issue. Adding only `GSUtil:parallel_process_count=1` without the thread count fixed it for me.

```
Google Cloud SDK 373.0.0
beta 2022.02.11
bq 2.0.73
core 2022.02.11
gsutil 5.6
```

**Comment by parity3** [comment]
It's hanging for me without the -m option at all. I ran out of local storage. Ideally it should have exited with an error. I won't bother creating a separate issue on that because I presume that handling out of space errors are probably in the realm "undetermined behavior / you're on your own".

**Comment by johanhaleby** [comment]
Revisiting this after many years, and unfortunately I run into the same problem even though I use:

> gsutil -o GSUtil:parallel_process_count=1 -o GSUtil:parallel_thread_count=24 -m cp -r gs://my-bucket .

I have Apple M1 Pro.

**Comment by andretw** [comment]
10k files and 100% reproducible. 99% Done forever. :'(

```
ation/json]...
- [10.0k/10.0k files][ 12.0 MiB/ 12.0 MiB]  99% Done  82.7 KiB/s ETA 00:00:00   
```

**Comment by layandreas** [comment]
Issue still persists.

So this ticket has been open for 6 years now. Does anyone actually maintain gsutil? If it's not fixed (/fixable) at least multiprocessing should be disabled by default on Mac.

**Comment by vlobanov** [comment]
@layandreas doesn't seem like gsutil is supported anymore. Use `gcloud storage` as @dilipped suggested above, I just tried and it apparently has multiprocessing upload enabled by default

**Comment by vasi-grigo** [comment]
i have not given it a proper test - but seems like one of the cases when it does hang is when the disk it is syncing to is out of space

**Comment by ste-wun** [comment]
> i have not given it a proper test - but seems like one of the cases when it does hang is when the disk it is syncing to is out of space

Can confirm - was puzzled for an hour what is happening. In the end the disk was full.
